Client/Frontend/UserContent/UserScripts/Scripts_Dynamic/ScriptHandlers/Paged/CosmeticFiltersScriptHandler.swift

```diff
@@ -52,21 +52,25 @@ class CosmeticFiltersScriptHandler: TabContentScript {
         return
       }
       
-      let domain = Domain.getOrCreate(forUrl: frameURL, persistent: tab?.isPrivate == true ? false : true)
-      let selectors = AdBlockStats.shared.cachedEngines(for: domain).flatMap { cachedEngine -> [String] in
-        do {
-          return try cachedEngine.selectorsForCosmeticRules(
-            frameURL: frameURL,
-            ids: dto.data.ids,
-            classes: dto.data.classes
-          )
-        } catch {
-          Logger.module.error("\(error.localizedDescription)")
-          return []
+      Task { @MainActor in
+        let domain = Domain.getOrCreate(forUrl: frameURL, persistent: tab?.isPrivate == true ? false : true)
+        let cachedEngines = AdBlockStats.shared.cachedEngines(for: domain)
+        
+        let selectorArrays = await cachedEngines.asyncConcurrentMap { cachedEngine -> [String] in
+          do {
+            return try cachedEngine.selectorsForCosmeticRules(
+              frameURL: frameURL,
+              ids: dto.data.ids,
```

Comment:
This might be a core data multithreading violation.
Domain is on a different thread than the concurrent map callback.

Reply:
I don't see which part of the code you are referencing here. There is no `Domain` usage in this concurrent map.

Reply:
If you mean this code `AdBlockStats.shared.cachedEngines(for: domain)` I do ensure the domain is on the main thread wherever I use it